### PR TITLE
Enable linker substitutions for more architectures.

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

**Review by jkotas — APPROVED** [review]
Thanks

**Comment by eerhardt** [comment]
```
  Value does not fall within the expected range.
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20459.8\tools\SymStore.targets(70,5): error MSB3073: The command ""F:\workspace\_work\1\s\.packages\microsoft.diasymreader.pdb2pdb\1.1.0-beta2-19575-01\tools\Pdb2Pdb.exe" "F:\workspace\_work\1\s\artifacts\bin\coreclr\Windows_NT.x64.Release\IL\System.Private.CoreLib.dll" /out "F:\workspace\_work\1\s\artifacts\SymStore\Release\System.Private.CoreLib\net5.0\x64\System.Private.CoreLib.pdb" /srcsvrvar SRC_INDEX=public" exited with code 2. [F:\workspace\_work\1\s\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
```

This looks like a new problem in the same area that we hit [the last time we tried this](https://github.com/dotnet/runtime/pull/37615#issuecomment-641493863).

cc @vitek-karas

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems [hunk 1]
```diff
@@ -28,11 +28,11 @@
...
-    <ILLinkSubstitutionsXmls Include="$(ILLinkSharedDirectory)ILLink.Substitutions.NoArmIntrinsics.xml" Condition="'$(Platform)' == 'wasm'" />
-    <ILLinkSubstitutionsXmls Include="$(ILLinkSharedDirectory)ILLink.Substitutions.NoX86Intrinsics.xml" Condition="'$(Platform)' == 'wasm'" />
+    <ILLinkSubstitutionsXmls Include="$(ILLinkSharedDirectory)ILLink.Substitutions.LittleEndian.xml" Condition="'$(Platform)' == 'wasm' or '$(Platform)' == 'arm' or '$(Platform)' == 'arm64' or '$(Platform)' == 'x86' or '$(Platform)' == 'x64'" />
+    <ILLinkSubstitutionsXmls Include="$(ILLinkSharedDirectory)ILLink.Substitutions.32bit.xml" Condition="'$(Platform)' == 'wasm' or '$(Platform)' == 'arm' or '$(Platform)' == 'x86'" />
+    <ILLinkSubstitutionsXmls Include="$(ILLinkSharedDirectory)ILLink.Substitutions.64bit.xml" Condition="'$(Platform)' == 'arm64' or '$(Platform)' == 'x64'" />
+    <ILLinkSubstitutionsXmls Include="$(ILLinkSharedDirectory)ILLink.Substitutions.NoArmIntrinsics.xml" Condition="'$(Platform)' == 'wasm' or '$(Platform)' == 'x86' or '$(Platform)' == 'x64'" />
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:34]: These conditions can be changed to use the above properties, to limit duplication.
```suggestion
    <ILLinkSubstitutionsXmls Include="$(ILLinkSharedDirectory)ILLink.Substitutions.NoArmIntrinsics.xml" Condition="'$(SupportsArmIntrinsics)' != 'true'" />
```

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems [hunk 2]
```diff
@@ -28,11 +28,11 @@
...
+    <ILLinkSubstitutionsXmls Include="$(ILLinkSharedDirectory)ILLink.Substitutions.NoX86Intrinsics.xml" Condition="'$(Platform)' == 'wasm' or '$(Platform)' == 'arm' or '$(Platform)' == 'arm64'" />
```
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:35]:
```suggestion
    <ILLinkSubstitutionsXmls Include="$(ILLinkSharedDirectory)ILLink.Substitutions.NoX86Intrinsics.xml" Condition="'$(SupportsX86Intrinsics)' != 'true'" />
```

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems (hunk 1 above)
Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:34]: Also note that the `arm` platform doesn't support Arm Intrinsics (yet).

**Comment by eerhardt** [comment]
This PR should resolve #31785

**Comment by jkotas** [comment]
> This PR should resolve #31785

It will resolve it for CoreLib. Ideally, we would do this for everything in the shared framework.

**Comment by eerhardt** [comment]
> Ideally, we would do this for everything in the shared framework.

That isn't possible without addressing the issues brought up in https://github.com/dotnet/runtime/issues/31712. If we need to do this for everything in the shared framework, we should re-open #31712 and address the issues brought up there.

**Comment by vitek-karas** [comment]
The PDB2PDB failure needs investigation - I'll get to it eventually, but given this is 6.0 only, I have to prioritize 5.0 work still.

**Comment by vitek-karas** [comment]
I was finally able to get back to this - the problem doesn't repro locally if I use the latest mono/linker master build of linker - so maybe we just need to pick up the latest linker in the runtime repo.

**Comment by eerhardt** [comment]
> so maybe we just need to pick up the latest linker in the runtime repo.

#43181 just brought in a new linker a few hours ago. Closing and re-opening this PR to re-run CI on the latest.

**Comment by eerhardt** [comment]
@vitek-karas - Looks like the CI is still failing even with the latest linker changes:

```
2020-10-09T19:11:52.2092988Z   Value does not fall within the expected range.
2020-10-09T19:11:52.2314176Z F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20506.7\tools\SymStore.targets(70,5): error MSB3073: The command ""F:\workspace\_work\1\s\.packages\microsoft.diasymreader.pdb2pdb\1.1.0-beta2-19575-01\tools\Pdb2Pdb.exe" "F:\workspace\_work\1\s\artifacts\bin\mono\Windows_NT.x64.Release\IL\System.Private.CoreLib.dll" /out "F:\workspace\_work\1\s\artifacts\SymStore\Release\System.Private.CoreLib\net6.0\x64\System.Private.CoreLib.pdb" /srcsvrvar SRC_INDEX=public" exited with code 2. [F:\workspace\_work\1\s\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
2020-10-09T19:11:52.2433282Z ##[error].packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20506.7\tools\SymStore.targets(70,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command ""F:\workspace\_work\1\s\.packages\microsoft.diasymreader.pdb2pdb\1.1.0-beta2-19575-01\tools\Pdb2Pdb.exe" "F:\workspace\_work\1\s\artifacts\bin\mono\Windows_NT.x64.Release\IL\System.Private.CoreLib.dll" /out "F:\workspace\_work\1\s\artifacts\SymStore\Release\System.Private.CoreLib\net6.0\x64\System.Private.CoreLib.pdb" /srcsvrvar SRC_INDEX=public" exited with code 2.
2020-10-09T19:11:52.2450431Z 
2020-10-09T19:11:52.2451041Z Build FAILED.
```

Using the `binlog` from the CI, I was able to verify that a recent build of the linker was used:

```
F:\workspace\_work\1\s\.dotnet\dotnet.exe "F:\workspace\_work\1\s\.packages\microsoft.net.illink.tasks\6.0.0-alpha.1.20509.3\tools\net5.0\illink.dll" -reference "F:\workspace\_work\1\s\artifacts\obj\coreclr\System.Private.CoreLib\x64\Release\PreTrim/System.Private.CoreLib.dll"
-out "F:\workspace\_work\1\s\artifacts\obj\coreclr\System.Private.CoreLib\x64\Release"
--warnaserror-  -t --strip-link-attributes false --ignore-link-attributes true --skip-unresolved true --disable-opt unusedinterfaces -r System.Private.CoreLib -c skip -u skip -p link System.Private.CoreLib -b true --strip-descriptors false --strip-substitutions false --keep-dep-attributes true --nowarn IL2008;IL2009;IL2012;IL2025;IL2026;IL2035;IL2050;IL2032;IL2055;IL2057;IL2058;IL2059;IL2060;IL2061;IL2062;IL2063;IL2064;IL2065;IL2066;IL2067;IL2068;IL2069;IL2070;IL2071;IL2072;IL2073;IL2074;IL2075;IL2076;IL2077;IL2078;IL2079;IL2080;IL2081;IL2082;IL2083;IL2084;IL2085;IL2086;IL2087;IL2088;IL2089;IL2090;IL2091
```

**Pull request merged**